README.md
# crtview - Terminal-based user interface toolkit

This project is a fork of [cview](https://gitlab.com/tslocum/cview),
which is also fork of original
[tview](https://github.com/rivo/tview). :-) See
``docs/about-crtview.md`` for more details.

## Features

Available widgets:

- __Input forms__ (including __input/password fields__, __drop-down selections__, __checkboxes__, and __buttons__)
- Navigable multi-color __text views__
- Selectable __lists__ with __context menus__
- Modal __dialogs__
- Horizontal and vertical __progress bars__
- __Grid__, __Flexbox__ and __tabbed panel layouts__
- Sophisticated navigable __table views__
- Flexible __tree views__
- Draggable and resizable __windows__
- An __application__ wrapper

Widgets may be customized and extended to suit any application.

## Installation

```bash
go get github.com/isbm/crtview
```

## Hello World

This basic example creates a TextView titled "Hello, World!" and displays it in your terminal:

```go
package main

import (
	"github.com/isbm/crtview"
)

func main() {
	app := crtview.NewApplication()
	
	box := crtview.NewTextView()
		.SetBorder(true)
		.SetTitle("Hello, world!")
		.SetText("Here is some meaning-less text for your app.")
	
	app.SetRoot(box, true)
	if err := app.Run(); err != nil {
		panic(err)
	}
}
```


CHANGELOG
v2.0.0 (2020-10-21)
- Fork cview as crtview
- Update docs, fork explanation etc
- GetBorder renamed to HasBorder
- Windows can have a status bar (dynamic text with tags)

v1.5.1 (2020-10-21 closed WIP, broken)
- Add FocusManager
- Add Slider
- Add TabbedPanels
- Add Application.GetScreen and Application.GetScreenSize
- Add SetVisible and GetVisible to all widgets
- Add TextView.SetBytes and TextView.GetBytes
- Add TableCell.SetBytes, TableCell.GetBytes and TableCell.GetText
- Fix List.Transform not calling handler set via SetChangedFunc
- Fix WordWrap bounds out of range
- Fix List dividers allowing selection
- Allow modification of scroll bar render text
- Allow scrolling List horizontally
- Clarify that Table rows must each have the same number of columns
- Generalize tag stripping as StripTags
- Make printWithStyle public and rename as PrintStyle
- Optimize TextView (writing is 90% faster, drawing is 50% faster)
- Remove return values from methods which return their primitive (breaks chaining)
- Remove Application.ForceDraw (Application.Draw may be called anywhere)
- Rename Pages as Panels
- Support bracketed paste mode via tcell

v1.5.0 (2020-10-03)
- Add scroll bar to TextView
- Add Window and WindowManager
- Add focus-driven style options
- Add InputField autocomplete style options
- Add arrow symbol to DropDown
- Add Makefile
- Allow autocomplete selection value customization
- Provide DropDownOption in DropDown handlers
- Provide ListItem in List handlers
- Panic when attempting to add an invalid FormItem
- Release lock on InputField while executing DoneFunc and FinishedFunc

v1.4.9 (2020-09-08)
- Add InputField.GetCursorPosition and InputField.SetCursorPosition
- Add Table.Sort, Table.SetSortFunc and Table.SetSortClicked
- Add TextView.SetReindexBuffer
- Base List PageUp and PageDown on the number of items displayed
- Upgrade tcell to v2: includes strikethrough support

v1.4.8 (2020-08-11)
- Add italic text formatting flag
- Add Modal.GetForm and Modal.GetFrame
- Fix Form.Clear deadlock
- Fill nil Flex space with default background color
- Position ContextMenu on selected item when negative coordinates are provided
- Use sync.RWMutex in all widgets

v1.4.7 (2020-06-09)
- Add Box.SetBackgroundTransparent
- Add List.SetSelectedAlwaysCentered
- Fix default background transparency of Flex and Grid
- Fix negative List offset and DropDown selection
- Fix panic when clicking nil primitive in Flex
- Fix deadlock when calling Pages.SendToFront
- Fix ANSI colors 0-15
- Fix ANSI SGR codes

v1.4.6 (2020-05-18)
- Add Box.ShowFocus
- Add Keys to allow default keyboard shortcuts to be modified
- Add List.GetOffset, List.SetOffset and List.SetSelectedTextAttributes
- Add TextView.SetMaxLines
- Add Vim-style keybindings to List
- Fix List not updating selected item before calling selected handlers
- Do not handle right click on List when there is no context menu
- Always initialize context menu List
- Document how to override default keybindings

v1.4.5 (2020-04-25)
- Add multithreading support
- Add ContextMenu (initially supported by List)
- Add List.Transform and TreeView.Transform
- Add Application.SetBeforeFocusFunc and Application.SetAfterFocusFunc
- Merge upstream mouse support

v1.4.4 (2020-02-24)
- Fix panic when navigating empty list
- Fix resize event dimensions on Windows
- Clarify that Box does not have inner text
- Do not wrap around form by default

v1.4.3 (2020-02-13)
- Add SetFocusedFunc to TreeNode
- Add option to always show scroll bar
- Fix scrolling Table with PageDown and PageUp
- Do not wrap around List by default

v1.4.2 (2020-02-02)
- Add scroll bar to List, DropDown, Table and TreeView
- Add SetDoneFunc to TreeView
- Fix mouse support when resuming after Application.Suspend
- Fix rendering issues with TextViews which have their background color set to
ColorDefault

v1.4.1 (2020-01-23)
- Add ProgressBar widget
- Add Application.RingBell
- Add example multi-layout application and mouse-enabled application
- Add documentation for GetChildren methods

v1.4.0 (2020-01-16)
- Bump version to resolve issues with "go get"

v0.2.2 (2020-01-06)
- Add optional message displayed after CheckBox
- Fix Dropdown mouse capture behavior
- Fix TextView region highlighting on last line

v0.2.1 (2020-01-01)
- Add initial mouse support (some widgets are unsupported)
- Add window size change handler
- Fix resetting TextView color with [-]
- Fix adding unnecessary padding to some runes
- Fix handling ANSI color escape codes 39 (default foreground) and 49 (default
background)

v0.2.0 (2019-12-30)
- Fork tview as cview

v0.1.20 (2019-07-08)
- Added autocomplete functionality to InputField.

v0.1.19 (2018-10-28)
- Added QueueUpdate() and QueueEvent() to Application to help with
modifications to primitives from goroutines.

v0.1.18 (2018-10-18)
- InputField elements can now be navigated freely.

v0.1.17 (2018-06-20)
- Added TreeView.

v0.1.15 (2018-05-02)
- Flex and Grid don't clear their background per default, thus allowing for
custom modals.

v0.1.14 (2018-04-13)
- Added an Escape() function which keep strings like color or region tags from
being recognized as such.
- Added ANSIWriter() and TranslateANSI() which convert ANSI escape sequences to
tview color tags.

v0.1.13 (2018-04-01)
- Added background colors and text attributes to color tags.

v0.1.12 (2018-03-13)
- Added "suspended mode" to Application.

v0.1.11 (2018-03-02)
- Added a RemoveItem() function to Grid and Flex.

v0.1.10 (2018-02-22)
- Direct access to the screen object through callback in Box (i.e. for all
primitives).

v0.1.9 (2018-02-20)
- Introduced Grid layout.
- Direct access to the screen object through callbacks in Application.

v0.1.8 (2018-01-17)
- Color tags can now be used almost everywhere.

v0.1.7 (2018-01-16)
- Forms can now also have a horizontal layout.

v0.1.6 (2018-01-14)
- All primitives can now intercept all key events when they have focus.

ansi.go
package crtview

import (
	"bytes"
	"fmt"
	"io"
	"strconv"
	"strings"
)

// The states of the ANSI escape code parser.
const (
	ansiText = iota
	ansiEscape
	ansiSubstring
	ansiControlSequence
)

// ansi is a io.Writer which translates ANSI escape codes into crtview color
// tags.
type ansi struct {
	io.Writer

	// Reusable buffers.
	buffer                        *bytes.Buffer // The entire output text of one Write().
	csiParameter, csiIntermediate *bytes.Buffer // Partial CSI strings.
	attributes                    string        // The buffer's current text attributes (a tview attribute string).

	// The current state of the parser. One of the ansi constants.
	state int
}

// ANSIWriter returns an io.Writer which translates any ANSI escape codes
// written to it into crtview color tags. Other escape codes don't have an effect
// and are simply removed. The translated text is written to the provided
// writer.
func ANSIWriter(writer io.Writer) io.Writer {
	return &ansi{
		Writer:          writer,
		buffer:          new(bytes.Buffer),
		csiParameter:    new(bytes.Buffer),
		csiIntermediate: new(bytes.Buffer),
		state:           ansiText,
	}
}

// Write parses the given text as a string of runes, translates ANSI escape
// codes to color tags and writes them to the output writer.
func (a *ansi) Write(text []byte) (int, error) {
	defer func() {
		a.buffer.Reset()
	}()

	for _, r := range string(text) {
		switch a.state {

		// We just entered an escape sequence.
		case ansiEscape:
			switch r {
			case '[': // Control Sequence Introducer.
				a.csiParameter.Reset()
				a.csiIntermediate.Reset()
				a.state = ansiControlSequence
			case 'c': // Reset.
				fmt.Fprint(a.buffer, "[-:-:-]")
				a.state = ansiText
			case 'P', ']', 'X', '^', '_': // Substrings and commands.
				a.state = ansiSubstring
			default: // Ignore.
				a.state = ansiText
			}

		// CSI Sequences.
		case ansiControlSequence:
			switch {
			case r >= 0x30 && r <= 0x3f: // Parameter bytes.
				if _, err := a.csiParameter.WriteRune(r); err != nil {
					return 0, err
				}
			case r >= 0x20 && r <= 0x2f: // Intermediate bytes.
				if _, err := a.csiIntermediate.WriteRune(r); err != nil {
					return 0, err
				}
			case r >= 0x40 && r <= 0x7e: // Final byte.
				switch r {
				case 'E': // Next line.
					count, _ := strconv.Atoi(a.csiParameter.String())
					if count == 0 {
						count = 1
					}
					fmt.Fprint(a.buffer, strings.Repeat("\n", count))
				case 'm': // Select Graphic Rendition.
					var background, foreground string
					params := a.csiParameter.String()
					fields := strings.Split(params, ";")
					if len(params) == 0 || len(fields) == 1 && fields[0] == "0" {
						// Reset.
						a.attributes = ""
						if _, err := a.buffer.WriteString("[-:-:-]"); err != nil {
							return 0, err
						}
						break
					}
					lookupColor := func(colorNumber int) string {
						if colorNumber < 0 || colorNumber > 15 {
							return "black"
						}
						return []string{
							"black",
							"maroon",
							"green",
							"olive",
							"navy",
							"purple",
							"teal",
							"silver",
							"gray",
							"red",
							"lime",
							"yellow",
							"blue",
							"fuchsia",
							"aqua",
							"white",
						}[colorNumber]
					}
				FieldLoop:
					for index, field := range fields {
						switch field {
						case "1", "01":
							if !strings.ContainsRune(a.attributes, 'b') {
								a.attributes += "b"
							}
						case "2", "02":
							if !strings.ContainsRune(a.attributes, 'd') {
								a.attributes += "d"
							}
						case "4", "04":
							if !strings.ContainsRune(a.attributes, 'u') {
								a.attributes += "u"
							}
						case "5", "05":
							if !strings.ContainsRune(a.attributes, 'l') {
								a.attributes += "l"
							}
						case "22":
							if i := strings.IndexRune(a.attributes, 'b'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
							if i := strings.IndexRune(a.attributes, 'd'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
						case "24":
							if i := strings.IndexRune(a.attributes, 'u'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
						case "25":
							if i := strings.IndexRune(a.attributes, 'l'); i >= 0 {
								a.attributes = a.attributes[:i] + a.attributes[i+1:]
							}
						case "30", "31", "32", "33", "34", "35", "36", "37":
							colorNumber, _ := strconv.Atoi(field)
							foreground = lookupColor(colorNumber - 30)
						case "39":
							foreground = "-"
						case "40", "41", "42", "43", "44", "45", "46", "47":
							colorNumber, _ := strconv.Atoi(field)
							background = lookupColor(colorNumber - 40)
						case "49":
							background = "-"
						case "90", "91", "92", "93", "94", "95", "96", "97":
							colorNumber, _ := strconv.Atoi(field)
							foreground = lookupColor(colorNumber - 82)
						case "100", "101", "102", "103", "104", "105", "106", "107":
							colorNumber, _ := strconv.Atoi(field)
							background = lookupColor(colorNumber - 92)
						case "38", "48":
							var color string
							if len(fields) > index+1 {
								if fields[index+1] == "5" && len(fields) > index+2 { // 8-bit colors.
									colorNumber, _ := strconv.Atoi(fields[index+2])
									if colorNumber <= 15 {
										color = lookupColor(colorNumber)
									} else if colorNumber <= 231 {
										red := (colorNumber - 16) / 36
										green := ((colorNumber - 16) / 6) % 6
										blue := (colorNumber - 16) % 6
										color = fmt.Sprintf("#%02x%02x%02x", 255*red/5, 255*green/5, 255*blue/5)
									} else if colorNumber <= 255 {
										grey := 255 * (colorNumber - 232) / 23
										color = fmt.Sprintf("#%02x%02x%02x", grey, grey, grey)
									}
								} else if fields[index+1] == "2" && len(fields) > index+4 { // 24-bit colors.
									red, _ := strconv.Atoi(fields[index+2])
									green, _ := strconv.Atoi(fields[index+3])
									blue, _ := strconv.Atoi(fields[index+4])
									color = fmt.Sprintf("#%02x%02x%02x", red, green, blue)
								}
							}
							if len(color) > 0 {
								if field == "38" {
									foreground = color
								} else {
									background = color
								}
							}
							break FieldLoop
						}
					}
					var colon string
					if len(a.attributes) > 0 {
						colon = ":"
					}
					if len(foreground) > 0 || len(background) > 0 || len(a.attributes) > 0 {
						fmt.Fprintf(a.buffer, "[%s:%s%s%s]", foreground, background, colon, a.attributes)
					}
				}
				a.state = ansiText
			default: // Undefined byte.
				a.state = ansiText // Abort CSI.
			}

			// We just entered a substring/command sequence.
		case ansiSubstring:
			if r == 27 { // Most likely the end of the substring.
				a.state = ansiEscape
			} // Ignore all other characters.

			// "ansiText" and all others.
		default:
			if r == 27 {
				// This is the start of an escape sequence.
				a.state = ansiEscape
			} else {
				// Just a regular rune. Send to buffer.
				if _, err := a.buffer.WriteRune(r); err != nil {
					return 0, err
				}
			}
		}
	}

	// Write buffer to target writer.
	n, err := a.buffer.WriteTo(a.Writer)
	if err != nil {
		return int(n), err
	}
	return len(text), nil
}

// TranslateANSI replaces ANSI escape sequences found in the provided string
// with crtview's color tags and returns the resulting string.
func TranslateANSI(text string) string {
	var buffer bytes.Buffer
	writer := ANSIWriter(&buffer)
	if _, err := writer.Write([]byte(text)); err != nil {
		panic("Error writing ANSI buffer: " + err.Error())
	}
	return buffer.String()
}
